pkg/image/controller/controller.go

```diff
@@ -35,243 +66,59 @@ const retryCount = 2
-// 3. image retrieving when error is different from RepositoryNotFound, RegistryNotFound or ImageNotFound
-// 4. ImageStreamMapping save error
-// 5. error when marking ImageStream as imported
+//
+// 3. spec.DockerImageRepository not defined - import tags per each definition.
+//
 func (c *ImportController) Next(stream *api.ImageStream) error {
-	if !needsImport(stream) {
+	ok, partial := needsImport(stream)
+	if !ok {
 		return nil
 	}
-	glog.V(4).Infof("Importing stream %s/%s...", stream.Namespace, stream.Name)
+	glog.V(3).Infof("Importing stream %s/%s partial=%t...", stream.Namespace, stream.Name, partial)
 
-	insecure := stream.Annotations[api.InsecureRepositoryAnnotation] == "true"
-	client := c.client
-	if client == nil {
-		client = dockerregistry.NewClient(5 * time.Second)
-	}
-
-	var errlist []error
-	toImport, retry, err := getTags(stream, client, insecure)
-	// return here, only if there is an error and nothing to import
-	if err != nil && len(toImport) == 0 {
-		if retry {
-			return err
-		}
-		return c.done(stream, err.Error(), retryCount)
-	}
-	if err != nil {
-		errlist = append(errlist, err)
-	}
-
-	retry, err = c.importTags(stream, toImport, client, insecure)
-	if err != nil {
-		if retry {
-			return err
-		}
-		errlist = append(errlist, err)
-	}
-
-	if len(errlist) > 0 {
-		return c.done(stream, kerrors.NewAggregate(errlist).Error(), retryCount)
+	isi := &api.ImageStreamImport{
+		ObjectMeta: kapi.ObjectMeta{
+			Name:            stream.Name,
+			Namespace:       stream.Namespace,
+			ResourceVersion: stream.ResourceVersion,
+		},
+		Spec: api.ImageStreamImportSpec{Import: true},
 	}
-
-	return c.done(stream, "", retryCount)
-}
-
-// getTags returns a map of tags to be imported, a flag saying if we should retry
-// imports, meaning not setting the import annotation and an error if one occurs.
-// Tags explicitly defined will overwrite those from default upstream image repository.
-func getTags(stream *api.ImageStream, client dockerregistry.Client, insecure bool) (map[string]api.DockerImageReference, bool, error) {
-	imports := make(map[string]api.DockerImageReference)
-	references := sets.NewString()
-
-	// read explicitly defined tags
-	for tagName, specTag := range stream.Spec.Tags {
-		if specTag.From == nil {
+	for tag, tagRef := range stream.Spec.Tags {
+		if !(partial && tagImportable(tagRef)) && !tagNeedsImport(stream, tag, tagRef, true) {
 			continue
 		}
-		if specTag.From.Kind != "DockerImage" || specTag.Reference {
-			references.Insert(tagName)
-			continue
-		}
-		ref, err := api.ParseDockerImageReference(specTag.From.Name)
-		if err != nil {
-			glog.V(2).Infof("error parsing DockerImage %s: %v", specTag.From.Name, err)
-			continue
+		isi.Spec.Images = append(isi.Spec.Images, api.ImageImportSpec{
+			From:         kapi.ObjectReference{Kind: "DockerImage", Name: tagRef.From.Name},
+			To:           &kapi.LocalObjectReference{Name: tag},
+			ImportPolicy: tagRef.ImportPolicy,
```

Comment:
New API takes priority - insecure was only a bulk policy, this is a specific policy on the tag that has higher priority than the bulk policy.  Old clients and servers can't import specific tags.